### DISCUSSION TITLE
Understanding how RAUC 1.10 detects file compression when updating an image

### DISCUSSION THREAD

**Discussion by mseeber** [rauc/rauc#1203] [Q&A] [answered]
In my setup i am using meta-rauc and have A and B slots containing an ext4 image for the rootfs and a `boot.tar.gz` for kernel, and devicetree.

When installing an update from a bundle, rauc attempts to unpack a boot image `boot.tar.gz.img`  to a vfat boot partition in the respective slot.

This has worked fine for me until the default provider for "tar" in my image switched from busybox to GNU tar.
In this case the update fails since rauc does not pass the -z flag to tar to decompress the file `boot.tar.gz.img`.

So in my case when looking at the code and the last commit that changed behavior of the respective function:
commit: https://github.com/rauc/rauc/commit/23b96a1f272510f70b524a5eae2d067567bf246f
code: https://github.com/rauc/rauc/blob/966a86b63b41be6cc677ae2fc8567a335efd511b/src/update_handler.c#L1364

This looks to me as if it was pure luck for me up to now, that this did work at all since busybox tar does not care about the missing `-z` flag when taking a compressed archive from stdin and did uncompress the archive even without the flag.

I would assume, that renaming the `boot.tar.gz.img` to `boot.tar.gz` inside the bundle would solve the issue, as `suffix_to_tar_flag` would match to the `.gz` suffix and provide the `-z` flag. The problem is, that the .img suffix in the bundle is added by the `bundle.bbclass` in the meta layer:

https://github.com/rauc/meta-rauc/blob/kirkstone/classes/bundle.bbclass#L272

Excerpt from the bundle recipe i use:

```
RAUC_SLOT_boot ?= "${PREFERRED_PROVIDER_virtual/kernel}"
RAUC_SLOT_boot[type] ?= "file"
RAUC_SLOT_boot[file] ?= "boot.tar.gz"
```

Should i select another type instead of `"file"` to avoid the `.img` suffix, or is there a better or more correct approach to this?

**Comment by ejoerns** [comment]
> This looks to me as if it was pure luck for me up to now, that this did work at all since busybox tar does not care about the missing -z flag when taking a compressed archive from stdin and did uncompress the archive even without the flag.

Not pure luck, but the documented required busybox config option `CONFIG_FEATURE_TAR_AUTODETECT=y`. :wink: 

You are right about the file name suffix. A `tar.gz` would be detected correctly.

Maybe one step back: I am a bit confused that the kernel is packaged into an archive. Is there more in the archive it than just the kernel?

**Reply by mseeber** [reply]
On the topic of CONFIG_FEATURE_TAR_AUTODETECT:

Documentation states that users may either use GNU tar or busybox tar. In case of busybox, the flags for autodetection need to be activated. That sounds reasonable to me but in my case the problem happens with with GNU tar, since GNU tar does not seem to support autodetection from stdin at all.

This is probably the reason the `suffix_to_tar_flag` function exists in that formanyway, not to support busybox but to support GNU tar.

So basically in the complete system, there are two layers to detect whether decompression is needed:

1. `suffix_to_tar_flag` adding the -z flag which is compatible with probably all implementations of tar
2. the specific implementation of tar may or may not detect compression and decompress the file even if 1. failed to detect compression.

So basically two layers of defense exist and both of them do not cover all cases right now. (but probably most reasonable cases, because my case admittedly is not the most straight forward one)

Would you be open to improve detection of the file type in RAUC by using MIME type detection, passing the correct flags more reliably, so the requirement on auto detection of the type in the tar implementation can be dropped?

**Reply by ejoerns** [reply]
The history is a bit different here than you assume. Before #811, all file names were directly passed to the tar process and all archives were detected by their file name extension. This worked for GNU tar anyway and for busybox tar with the above-mentioned config option enabled.

With #811, the handling needed to be changed to hook between the reading and the unpack process for generating progress data. Since then, for these update handlers only the manual detection implemented in `suffix_to_tar_flag` works. Neither busybox tar nor GNU tar should be able to guess the compression type from the stdin stream.

The advantage of using the file name suffix is that it is a quite lightweight implementation.
So using the right file name extensions should be the right solution here.

How would you implement MIME or file type detection?

**Reply by mseeber** [reply]
I imagine calling into the "file" binary could provide a mime type string that could be matched against, something like:

```
# file --mime-type - < boot.tar.gz.img                                                                                            
/dev/stdin: application/gzip
```
or otherwise one could call into a library to detect the MIME type.

Both options are less lightweight though and add a dependency which may not be desirable and may come with their own caveats.

I agree though, that a correct file name is the preferred solution.

**Reply by ejoerns** [reply]
Yes, each of your paragraphs exactly matches my recent thoughts on this topic :smirk:

**Comment by mseeber** [comment]
Yes, there are more files in the boot.tar.gz, mainly device tree and the bootenv for u-boot.

```
tar -tf boot.tar.gz.img
./
./tee.bin
./imx8mp-phycore-no-eth.dtbo
./imx8mp-amrum-fumu2-button-board-dev.dtbo
./imx8mp-phycore-no-spiflash.dtbo
./imx8mp-amrum-fumu2-imx477-csi1.dtbo
./imx8mp-phycore-rpmsg.dtbo
./oftree
./imx8mp-phyboard-pollux-peb-wlbt-05.dtbo
./Image
./imx8mp-amrum-fumu2-usb0p.dtbo
./imx8mp-phyboard-pollux-peb-av-010.dtbo
./imx8mp-phyboard-pollux-peb-av-012.dtbo
./bootenv.txt
./imx8mp-amrum-fumu2-interface-board-dev.dtbo
./imx8mp-amrum-fumu2-usb1p.dtbo
./imx8mp-phycore-no-rtc.dtbo
./imx8mp-amrum-fumu2-isp-csi1.dtbo
./imx8mp-amrum-fumu2-core.dtb
./imx8mp-amrum-fumu2-led-driver-board-dev.dtbo
```

The vendor layers for my SoM are doing a lot of interesting stuff i still have to dig through, maybe some info about the bundle helps to roughtly unerstand the setup:

```
Bundle Format:  verity                                                                                                                                      
  Verity Salt:  'a043c301a35c3a03984cd03cb109664328f118be04cb232b6e808499bc7022a2'                                                                          
  Verity Hash:  'e60869fd862d9386557aeeaccb29bddd1909689c1aa76f70a98b8788c1ea6dfc'                                                                          
  Verity Size:  4653056
Manifest Hash:  '20d5df5c9b57623902661bc9e247cf1de9bee3802f9ea96ac8888569eeca66ec'                                                                          

                                                                                                                                                            
3 Images:
  [rootfs]
        Filename:  amrum-vision-image-amrum-fumu2-imx8mp-3.ext4
        Checksum:  f17c2020ad6c8889bb9c3c2e527a6a99e2f0f62a5ae03136952ccb85d1ed1f75
        Size:      2.2?GB (2158772224 bytes)
        Hooks:
        Adaptive:  block-hash-index
  [boot]
        Filename:  boot.tar.gz.img
        Checksum:  d9d5f5c4f2adc3e31c6cd9ecd7e0be0db4db7521d2703aa12f4da8c875a333a1
        Size:      13.1?MB (13081488 bytes)
        Hooks:
  [bootloader]
        Filename:  emmc-boot.img
        Checksum:  ac8b775e75ee77e006a7da21a42bdbb64be9dde62180c6938334660af831e48a
        Size:      2.1?MB (2107264 bytes)
        Hooks:
```

On the target it looks like this:

```
=== Slot States ===
  [bootloader.0] (/dev/mmcblk2, boot-emmc, inactive)

o [rootfs.1] (/dev/mmcblk2p6, ext4, inactive)
        bootname: system1
        boot status: bad
    [boot.1] (/dev/mmcblk2p2, vfat, inactive)

x [rootfs.0] (/dev/mmcblk2p5, ext4, booted)
        bootname: system0
        mounted: /
        boot status: good
    [boot.0] (/dev/mmcblk2p1, vfat, active)
```

The journal tells me for a failed update:

```
# journalctl --no-pager -r -u rauc -n20
Apr 29 11:40:22 amrum-fumu2-imx8mp-3 rauc[6770]: installing `/root/amrum-vision-bundle-amrum-fumu2-imx8mp-3.raucb` failed: 1
Apr 29 11:40:22 amrum-fumu2-imx8mp-3 rauc[6770]: installing /root/amrum-vision-bundle-amrum-fumu2-imx8mp-3.raucb: finished
Apr 29 11:40:22 amrum-fumu2-imx8mp-3 rauc[6770]: installing /root/amrum-vision-bundle-amrum-fumu2-imx8mp-3.raucb: Installation error: Failed updating slot bo
ot.1: failed to splice data to tar: Broken pipe
Apr 29 11:40:22 amrum-fumu2-imx8mp-3 rauc[6770]: Installation error: Failed updating slot boot.1: failed to splice data to tar: Broken pipe
Apr 29 11:40:22 amrum-fumu2-imx8mp-3 rauc[6770]: Unmounting vfat slot /dev/mmcblk2p2
Apr 29 11:40:22 amrum-fumu2-imx8mp-3 rauc[6883]: tar: Error is not recoverable: exiting now
Apr 29 11:40:22 amrum-fumu2-imx8mp-3 rauc[6883]: tar: Archive is compressed. Use -z option
Apr 29 11:40:22 amrum-fumu2-imx8mp-3 rauc[6770]: Extracting /run/rauc/bundle/boot.tar.gz.img to /run/rauc/boot.1
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6770]: Mounting vfat slot /dev/mmcblk2p2
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6881]: mkfs.fat 4.2 (2021-01-31)
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6770]: Formatting vfat slot /dev/mmcblk2p2
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6770]: Updating /dev/mmcblk2p2 with /run/rauc/bundle/boot.tar.gz.img
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6770]: installing /root/amrum-vision-bundle-amrum-fumu2-imx8mp-3.raucb: Updating slot boot.1
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6770]: installing /root/amrum-vision-bundle-amrum-fumu2-imx8mp-3.raucb: Checking slot boot.1
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6770]: installing /root/amrum-vision-bundle-amrum-fumu2-imx8mp-3.raucb: Updating slot rootfs.1 done
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6770]: installing /root/amrum-vision-bundle-amrum-fumu2-imx8mp-3.raucb: Updating slot rootfs.1 status
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6880]: The filesystem on /dev/mmcblk2p6 is now 583984 (4k) blocks long.
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6880]: Resizing the filesystem on /dev/mmcblk2p6 to 583984 (4k) blocks.
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6880]: resize2fs 1.46.5 (30-Dec-2021)
Apr 29 11:40:21 amrum-fumu2-imx8mp-3 rauc[6770]: Resizing /dev/mmcblk2p6
```

So the idea of the  vendor seems to be to just dump the contents of the archive into a vfat partition which gets the used by u-boot to load the kernel image and the device tree + overlays listed in bootenv.txt .

I have searched through the meta layers and the only place the suffix is added seems to be in bundle.class in meta-rauc, which decides it needs to be there because the slot type is "file".

**Reply by jluebbe** [reply]
This only worked by accident, as `*.tar*` is matched before `*.img` (which would trigger raw writing of the file to the partition).

Tar archives compressed with gzip need to have the `.tar.gz` extension in the bundle, not `.img`. So for this case, the `bundle.bbclass` should not add a `.img` suffix.

**Reply by mseeber** [reply]
Exactly, the class does that though, because whether the suffix is added is determined by slot flags and image type, at least when looking at the function for writing the manifest.

I will dig through the bbclass further to see what is going on there.

**Reply by ejoerns** [reply, marked as answer]
Yes, for type `file`, the meta-rauc class unconditionally appends a `.img` to the file name. This was thought of as convenience for arbitrary image files packed into the bundle (like the `.dtb` device tree file from the classes example).
My assumption was that RAUC does not know most of the file extensions used there and that renaming them to .img to let RAUC just raw copy them to the target slot. Of course, this will fail with tar archives that should be extracted.

Tar archives we had so far mostly with rootfs images where no renaming will happen.
And for normal kernel images, we have the distinct 'kernel' type.
Both will not apply here.

I am not really sure what the best way to solve this would be. If we just remove the renaming, it might brake other's use cases that expect this behavior. Maybe adding a new type would work. But this makes the type handling even more complex...
I am also not fully convinced anymore if we created more convenience with this 'type' handling or added just another layer of confusion :smirk:

A workaround that should work without modifying the bundle class is to use the `[rename]` varflag:

```
RAUC_SLOT_boot[type] ?= "file"
RAUC_SLOT_boot[file] ?= "boot.tar.gz"
RAUC_SLOT_boot[rename] ?= "boot.tar.gz"
```

**Reply by mseeber** [reply]
Thanks a lot, i will take a look at the rename flag. If tat works i will let you know.

**Reply by mseeber** [reply]
The rename varflag has done the trick. Updates now continue to work as intended.
